### PR TITLE
Fix some HID bugs

### PR DESCRIPTION
**Pull request by gblues**
## Description

Getting the DS3 working again led me down an interesting rabbit hole.

For some reason, most of the HID implementations (save 2: wii u and bluetooth) do some funny business with the pad data buffer when reading input. Here's an example in the wiiusb HID driver:

```c
   if (hid && hid->connections && size > 0)
      pad_connection_packet(&hid->connections[adapter->slot],
            adapter->slot, adapter->data-1, size+1);
   // what the hell is this?        ^^^^^^^^^^^^^^
```
It doesn't make any sense. It can also blow up if adapter->data is a pointer and not an array reference. But it's present dating back 6 years ago when the wiiusb HID driver was first added. And there's no reason for it now.

Of course, (almost) all of the connect_* drivers expect this math to have been done, which means the packet offsets are all in fact off-by-one. The exceptions being .. the DS3 and the WiiU Gamecube Adapter.

It turns out that the `#ifdefs` I'd put in those drivers to handle Apple compatibility were in fact caused by this same error.

Change summary:
1. Take out the funny math done in affected HID drivers so we're not playing games with the adapter data pointer.
2. Remove the `#if (APPLE)` code from ps3 and GCA drivers
3. Go through each of the other connect_ hid pad drivers and subtract 1 from the offsets referenced
4. Revert a commit added a couple days ago that broke Mac OS HID
5. Add a new PID for a later-model DS4 pad (read: the ones I have lying around)

## Reviewers

@LibretroAdmin 

I've tested these changes on MacOS 12.6 (Monterey) with both a DS3 and DS4.

### CODE REVIEW

File: input/drivers_hid/iohidmanager_hid.c [hunk 1]
```diff
@@ -1105,7 +1105,7 @@ static int32_t iohidmanager_set_report(void *handle, uint8_t report_type, uint8_
       (struct iohidmanager_hid_adapter*)handle;
 
    if (adapter)
-      return IOHIDDeviceSetReport(adapter->handle, translate_hid_report_type(report_type), report_id, data_buf, size);
+      return IOHIDDeviceSetReport(adapter->handle, translate_hid_report_type(report_type), report_type, data_buf, size);
```
Comment (23rd) [inline input/drivers_hid/iohidmanager_hid.c:1108]: Do you think it was really designed that way?
The second argument `uint8_t report_type` takes the inner RetroArch's values, that the macOS has no idea about.
https://github.com/libretro/RetroArch/blob/0973e8d4b071d10602b25b5a2271bbec062dcf03/input/include/hid_driver.h#L25-L28

In that case, why is there `report_id` processing inside the `connect_ps3.c` file at all, if it is not used absolutely nowhere?

**Comment by 23rd** [comment]
I don't know exactly but perhaps the `funny math` was added cause of a signature of [IOHIDDeviceSetReport](https://developer.apple.com/documentation/iokit/1588656-iohiddevicesetreport).

> ID of the report being sent. **If the device supports multiple reports, this should also be set in the first byte of the report.**

As I can see, an implementation of HID manager in RetroArch does not have any checks for support of multiple reports, so in _some_ cases it makes sense, and in some cases it doesn't.

Look at these lines in a Chromium's implementation.
https://github.com/chromium/chromium/blob/d8cd0c3dcb69f76443b87ba3e7e93749b6117a78/services/device/hid/hid_connection_mac.cc#L163-L168

File: input/drivers_hid/iohidmanager_hid.c (hunk 1 above)
Comment (gblues) [reply input/drivers_hid/iohidmanager_hid.c]: No, I think I'm an idiot. I apologize. I've zapped that commit and done a proper fix for the DS3.

**Comment by gblues** [comment]
Hey @revvv, can you test this to see if the retrode still functions properly on wiiu?

**Comment by revvv** [comment]
> Hey @revvv, can you test this to see if the retrode still functions properly on wiiu?

With your fix [152543b](https://github.com/libretro/RetroArch/pull/14435/commits/152543bc000eb69feeb5d998334d5735fc81cc37) it will not work. None of my USB controllers.

Have you tested your changes on the Wii and Wii U? I guess if it works on Wii, it doesn't on Wii U and vice versa.
Does any other USB controller work on Wii U? I guess not, because all connect_*.c drivers expect the bytes at another position.
IMHO you have to change your connect_*.c. implementation. Just add +1 to packet[...+1].

**Comment by revvv** [comment]
@ gblues Sorry, I haven't seen all your changes. Have to test, but I really don't know when...

**Comment by LibretroAdmin** [comment]
@revvv hi, let us know when you can get down to look into this. We hope to be able to release our new stable before the end of the month.

**Comment by revvv** [comment]
> @revvv hi, let us know when you can get down to look into this. We hope to be able to release our new stable before the end of the month.

Reading the code it looks fine. I tried to test it, but my private build file seems broken.
Wii U crashes with "There is a problem with the system memory." Don't know what I have done wrong...

**Comment by revvv** [comment]
Just had the wrong devkitpro version... Now I could test it... unfortunately my controllers don't work without these fixes:
```
/input/connect/connect_nesusb.c
-   val = device->data[4 + axis];
+   val = device->data[3 + axis];

/input/connect/connect_retrode.c
-      val = pad->data[2 + axis];
+      val = pad->data[1 + axis];

-      val = device->pad_data[0].data[2 + axis];
+      val = device->pad_data[0].data[1 + axis];
```
I'm terribly sorry for this poor diff...

I guess the other controllers might also need to fix their get_axis() methods.

**Pull request merged**